connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala

```diff
@@ -0,0 +1,239 @@
+    SparkConnectArtifactManager.getOrCreateArtifactManager
+
+  override def onNext(req: AddArtifactsRequest): Unit = {
+    if (this.holder == null) {
+      this.holder = SparkConnectService.getOrCreateIsolatedSession(
+        req.getUserContext.getUserId,
+        req.getSessionId)
+    }
+
+    if (req.hasBeginChunk) {
+      // The beginning of a multi-chunk artifact.
+      require(chunkedArtifact == null)
+      chunkedArtifact = writeArtifactToFile(req.getBeginChunk)
+    } else if (req.hasChunk) {
+      // We are currently processing a multi-chunk artifact
+      require(chunkedArtifact != null && !chunkedArtifact.isFinished)
+      chunkedArtifact.write(req.getChunk)
+
+      if (chunkedArtifact.isFinished) {
+        chunkedArtifact.close()
+        // Unset the currently active chunked artifact.
+        chunkedArtifact = null
+      }
+    } else if (req.hasBatch) {
+      // Each artifact in the batch is single-chunked.
+      req.getBatch.getArtifactsList.forEach(artifact => writeArtifactToFile(artifact).close())
+    } else {
+      throw new UnsupportedOperationException(s"Unsupported data transfer request: $req")
+    }
+  }
+
+  override def onError(throwable: Throwable): Unit = {
+    Utils.deleteRecursively(stagingDir.toFile)
+    responseObserver.onError(throwable)
+  }
+
+  protected def addStagedArtifactToArtifactManager(artifact: StagedArtifact): Unit = {
+    artifactManager.addArtifact(holder.session, artifact.path, artifact.stagedPath)
+  }
+
+  /**
+   * Process all the staged artifacts built in this stream.
+   *
+   * @return
+   */
+  protected def flushStagedArtifacts(): Seq[ArtifactSummary] = {
+    // Non-lazy transformation when using Buffer.
+    stagedArtifacts.map { artifact =>
+      // We do not store artifacts that fail the CRC. The failure is reported in the artifact
+      // summary and it is up to the client to decide whether to retry sending the artifact.
+      if (artifact.getCrcStatus.contains(true)) {
+        addStagedArtifactToArtifactManager(artifact)
+      }
+      artifact.summary()
+    }.toSeq
+  }
+
+  protected def cleanUpStagedArtifacts(): Unit = Utils.deleteRecursively(stagingDir.toFile)
+
+  override def onCompleted(): Unit = {
+    val artifactSummaries = flushStagedArtifacts()
+    // Add the artifacts to the session.
+    val builder = proto.AddArtifactsResponse.newBuilder()
+    artifactSummaries.foreach(summary => builder.addArtifacts(summary))
+    // Delete temp dir
+    cleanUpStagedArtifacts()
+
+    // Send the summaries and close
+    responseObserver.onNext(builder.build())
+    responseObserver.onCompleted()
+  }
+
+  /**
+   * Create a (temporary) file for a single-chunk artifact.
+   */
+  private def writeArtifactToFile(
+      artifact: proto.AddArtifactsRequest.SingleChunkArtifact): StagedArtifact = {
+    val stagedDep = new StagedArtifact(artifact.getName)
+    stagedArtifacts += stagedDep
+    stagedDep.write(artifact.getData)
+    stagedDep
```

Comment:
You could close it right away. I am not sure what that brings you; it might be easier to reason about.

Reply:
The `close` here closes both the output stream and builds the artifact summary (essentially, its a "close" for the artifact rather than only the output stream) so there is a bit of logic separation between the "writing" side and the "closing" side (and hence, the write method doesn't take on the additional task).
It's also consistent for single-chunk and multi-chunk artifacts.